### DISCUSSION TITLE
Hooking into Spring component creation with new module

### DISCUSSION THREAD

**Discussion by rickfish** [Netflix/conductor#2698] [Q&A] [answered]
I am trying to move our modules from Conductor 2.31 to the new Spring model in 3.4.1. I am having difficulty figuring out how to get my components created by Spring. I have a 'properties' class with the @ConfigurationProperties annotation, a 'configuration' class with the @EnableConfiguriationProperties annotation pointing to that 'properties' class and a component that uses those properties in a Java class annotated with @Component. The 'configuration' class creates a @Bean for the component. When Conductor starts up, it does not automatically instantiate any of these classes.

I have reviewed a lot of the Spring documentation to try and figure out how it all fits together and I have looked at the other modules as examples but so far nothing has helped.

Two questions:
1. Is there any documentation that describes how to bootstrap a new module using the new Spring configuration?
2. Is there a place in the codebase that dictates the classpath that Spring uses to scan the classes?  I haven't found any @ComponentScan annotation anywhere...

**Comment by aravindanr** [comment, marked as answer]
> Is there any documentation that describes how to bootstrap a new module using the new Spring configuration?

There is none, since Conductor is like any other spring boot application. But this is the first request we had about someone moving modules from 2.31 to 3.x. We will consider adding a guide.

> Is there a place in the codebase that dictates the classpath that Spring uses to scan the classes? I haven't found any @componentscan annotation anywhere...

`componentscan` is not needed when `SpringBootApplication` is [used](https://github.com/Netflix/conductor/blob/main/server/src/main/java/com/netflix/conductor/Conductor.java#L29). By default all classes under `com.netflix.conductor` are scanned. My guess is that the new module you are creating doesn't follow that package name. 

If this new module is for contribution to Conductor, i recommend following the package name. 

If its for internal use and want to use a custom package name, i suggest changing the [SpringBootApplication](https://docs.spring.io/spring-boot/docs/2.0.x/reference/html/using-boot-using-springbootapplication-annotation.html) annotation like below,

`@SpringBootApplication(exclude = DataSourceAutoConfiguration.class, scanBasePackages = {"com.netflix.conductor", "<new module's top level package name here>"})`

Hope this helps. Let us know if you run into issues.

**Reply by rickfish** [reply]
Thanks @aravindanr, that is what I was looking for. I do have a different package name to differentiate my stuff since it is not ready to contribute.

**Comment by aravindanr** [comment]
@rickfish We would like to know more about Conductor users. As a start, we are gathering who is using Conductor. Appreciate if you can send a PR containing your company information in [WHOSUSING.md](https://github.com/Netflix/conductor/blob/main/WHOSUSING.md).

**Reply by rickfish** [reply]
Added my company info. Note that I am a contractor to that company.
